core/storage/statistics/gae_models.py

```diff
@@ -1134,6 +1138,132 @@ def delete_playthroughs_multi(cls, playthrough_ids):
+        # NOTE: In future if this model is used for any other entity type,
+        # then make the id_parameters dict for it, for eg. id_parameters
+        # dict for entity type exploration is {'exp_id': '', 'state_name': ''}
+        #  and add an elif condition in this function with the method to
+        #  generate the entity_id for that entity.
+        if entity_type == feconf.ENTITY_TYPE_EXPLORATION:
+            return '%s.%s' % (
+                id_parameters['exp_id'], id_parameters['state_name'])
+        else:
+            raise Exception('Invalid entity type.')
+
+    @classmethod
+    def generate_new_answer_details_id(cls, entity_type, entity_id):
+        """Generates a new answer detail ID which is unique.
+
+        Args:
+            entity_type: str. The type of the entity.
+            entity_id: str. The ID of the entity.
+
+        Returns:
+            str. An answer detail ID that is different from the IDs of all
+                the existing answer details within the given entity.
+
+        Raises:
+           Exception: There were too many collisions with existing answer
+               detail IDs when attempting to generate a new answer detail ID.
+        """
+        for _ in range(_MAX_RETRIES):
+            answer_details_id = (
+                entity_type + '.' + entity_id + '.' +
+                utils.base64_from_int(utils.get_current_time_in_millisecs()) +
+                utils.base64_from_int(utils.get_random_int(_RAND_RANGE)))
+            if not cls.get_by_id(answer_details_id):
+                return answer_details_id
+        raise Exception(
+            'New answer details id generator is producing too many collisions.')
+
+    @classmethod
+    def create(
+            cls, entity_id, entity_type, learner_answer,
+            learner_answer_details):
+        """Creates a new LearnerAnswerDetailsModel and
+        then writes it to the datastore.
+
+        Args:
+            entity_id: str. ID of the entity.
+            entity_type: str.  The type of entity, that whether it is state,
+                question, or any other object if it is added in future.
+            learner_answer: str. The answer entered by the user.
+            learner_answer_details: str. The details of the answer entered by
+                the learner.
+
+        Returns:
+            LearnerAnswerDetailsModel. The instance of the
+                LearnerAnswerDetailsModel.
+        """
+        answer_details_id = cls.generate_new_answer_details_id(
+            entity_type, entity_id)
+        answer_details_instance = cls(
+            id=answer_details_id,
+            entity_id=entity_id,
+            entity_type=entity_type,
+            learner_answer=learner_answer,
+            learner_answer_details=learner_answer_details,
+            created_on=datetime.datetime.utcnow())
+        answer_details_instance.put()
+        return answer_details_instance
+
+    @classmethod
+    def get_answer_details(
```

Comment:
Shouldn't you be fetching by state? Doesn't really make sense to see all the answer details for different states lumped together.

I'm a bit concerned that this PR suggests that that design consideration hasn't been taken into account; you may want to revisit the pipeline to make sure you're structuring your models for the right sorts of queries in the future. You should also have a plan for extending this to Questions, but I only see Explorations here.

Reply:
It will fetch it by state, as I am using entity_id here which will be in the form 'exploration_id'.'state_name'. So for a particular state, it will get all the answer details.

Also I tried making it generalized as I am using id_parameters:
The id parameters for the exploration will be ```{'exp_id':'the id of exp', 'state_name':'The name of state'}```, if the entity type is a question then there will be different id_parameters like ```{'question_id': 'the id of question', 'state_name': 'the name of the state'}```. There will be a get_id_parameters_for_exp function in the services file which will get the id_parameter for exp, similarly, we can add a function in the services which will generate id_parameter for any entity type like get_id_parameters_for_question. Thus as we will have the entity_type so we can check with if condition which id_parameters function needs to be called. And then we can pass id_parameters with entity_type to generate entity_id for any entity.
Do let me know if you have any doubt in this method?

Reply:
![WhatsApp Image 2019-06-05 at 7 35 51 AM](https://user-images.githubusercontent.com/30555544/58925492-56401e00-8765-11e9-9862-c4692105ba99.jpeg)

Reply:
Ah, thanks for explaining. We generally use "entity" to mean a datastore entity, so it's only at the level of "exploration", "question", "skill, etc. but not "state". So I think we need a different naming convention. How do we do it in the generalized review system? /cc @nithusha21 @DubeySandeep 

Also, I think you should keep the ID construction logic in the storage layer, since it is only relevant there and nowhere else (right)? Similar to what we do for some other models like stats. 

I also think you should be implementing specific functions with explicit args for question/exploration in this PR, rather than passing in dicts. Dicts are just data blobs which are hard to validate, and having separate args helps constrain things more and keeps the code more maintainable.

Reply:
We call it target_type for 'exploration', 'question', 'topic' etc and target_id for the id (target makes sense in terms of suggestions). 

Maybe you could call this the instance_id?

Reply:
Redesigned model doc in the PR explanation.